### PR TITLE
Polish configuration API

### LINKED ISSUE #394: Revisit Configuration API
There are a couple of inconsistencies in the configuration API.

The API should comply with this
* after creating the configuration from env it should be possible to override any value via programmatic API
* Configurations can be created via programmatic API - via constructor not only via `fromEnv`

* programmatic api does not allow to override set service name
* codec configuration can be created only via `.fromEnv()`
* most of the configurations provide getters but not for all properties

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #394 

- [x] there are getters for all configuration properties
- [x] codec configuration can be created programmatically
- [x] allow overriding service name

Configuration can be created programmatically o via env. If it is created via env all properties can be changed programmatically.

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

File: jaeger-core/src/main/java/io/jaegertracing/Configuration.java [hunk 1]
```diff
@@ -430,6 +447,14 @@ public SamplerConfiguration withManagerHostPort(String managerHostPort) {
   public static class CodecConfiguration {
     private Map<Format<?>, List<Codec<TextMap>>> codecs;
 
+    public CodecConfiguration() {
+    }
+
+    public CodecConfiguration withCodec(Format<?> format, Codec<TextMap> codec) {
```
Comment (aaronkorver) [inline jaeger-core/src/main/java/io/jaegertracing/Configuration.java]: Yay!!

**Comment by aaronkorver** [comment]
A note, not necessary on this PR as I can open a separate issue, the `withTags()` would be great if there was exposed a `withTagsFromEnv()` as well that could use the private method that is currently in the config.

**Comment by pavolloffay** [comment]
@aaronkorver you can call ` c = Configuration.fromEnv()` then `c.getTags()` and then `c.withTags(tags)` to fully manipulate with tags.

**Comment by aaronkorver** [comment]
right, but I can't call the parser function in Configuration as it is private right now.  I have some teams that want to take advantage of the `JAEGER_TAGS` environment.  My team has been providing a library for tracing and want them to use it as well, but have basically cut/paste that function into our library so we can configure certain things from the ENV, but override other pieces.  Hence the suggestion to do `withTagsFromEnv()`

**Comment by pavolloffay** [comment]
I am not sure whether method `c.withTagsFromEnv()` makes sense as the tags from env are obtained during `Configuration.fromEnv()`.

We could perhaps expose `map = Configuration.tracerTagsFromEnv()` , then you can get tags from env and programmatically mix things together. Oh wait actually right now you can call `fromEnv()` and get parsed tags from the configuration.

File: jaeger-core/src/main/java/io/jaegertracing/Configuration.java [hunk 2]
```diff
@@ -280,8 +280,14 @@ public void setStatsFactory(StatsFactory statsFactory) {
   /**
    * @param metricsFactory the MetricsFactory to use on the Tracer to be built
    */
-  public void withMetricsFactory(MetricsFactory metricsFactory) {
+  public Configuration withMetricsFactory(MetricsFactory metricsFactory) {
```
Comment (pavolloffay) [inline jaeger-core/src/main/java/io/jaegertracing/Configuration.java]: This shouldn't break clients

**Comment by aaronkorver** [comment]
So are you recommending that you should first do `myConfig = Configuration.fromEnv()` and then change things within `myConfig`?  The unfortunate part on that is the `getTracerTags()` returns an unmodifiableMap() that I can't merge things.

**Comment by pavolloffay** [comment]
> So are you recommending that you should first do myConfig = Configuration.fromEnv() and then change things within myConfig

Yes, the map is unmodifiable for a purpose, you can do `m = new HasMap<>(unmodifiableMap)`

File: jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java [hunk 3]
```diff
@@ -328,6 +333,14 @@ public void testNoServiceName() {
     new Configuration(null);
   }
 
+  @Test
+  public void testOverrideServiceName() {
```
Comment (jpkrohling) [inline jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: Can you add a test using `withServiceName(null)`, and another with `withServiceName("")`? This should document what is expected to happen in such case.

Comment (pavolloffay) [reply jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: It's not strictly related to this PR, but I am happy to add it. 

Test for null service name is already there, I have added only for empty case.

File: jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java [hunk 4]
```diff
@@ -328,6 +333,19 @@ public void testNoServiceName() {
     new Configuration(null);
   }
 
+  @Test(expected = RuntimeException.class)
+  public void testEmptyServiceName() {
+    new Configuration("");
```
Comment (jpkrohling) [inline jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: I actually thought about having the `null` overriding a valid existing name:

```java
new Configuration("Test").withServiceName(null);
```

This way, you'd be documenting the expected behavior for this scenario, as it might be ambiguous. This is not required, so, I'm giving my +1 to this PR.

Comment (pavolloffay) [reply jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: The check is also in tracer constructor, but in this case we wanted to fail before tracer-init time
